### PR TITLE
ci: add bazel_repository_cache to CircleCI workspace

### PR DESCRIPTION
**Pull request by filipesilva**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?
Rename `install` job to `setup` as requested in https://github.com/angular/angular/pull/28928#pullrequestreview-207023800

Restore `bazel_repository_cache` caching behaviour in CircleCI workspace usage.

PR target should be the same as https://github.com/angular/angular/pull/28928, but at the moment I am not sure if it's `master` or `master & patch`.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -136,16 +137,26 @@ jobs:
       - *init_environment
       - *yarn_install
       - run: yarn --cwd aio install --frozen-lockfile --non-interactive
+      # Make the bazel directories and add a file to them if they don't exist already so that
+      # persist_to_workspace not fail.
```
Comment (gkalpak) [inline .circleci/config.yml]: _not fail_ --> _does not fail_

**Review by gkalpak — CHANGES_REQUESTED** [review]
There are more occurrences of the `install` job that need to be changed to `setup`: https://github.com/angular/angular/blob/7acfa389a360a1856486ecec3e2550cbed780193/.circleci/config.yml#L637-L640

I am worried about two things with these type of changes:
- The persisted stuff can get quite large and costly (e.g. in terms of time). That might be less of an issue if workflow jobs can take advantage of locality (nit sure how CircleCI infrastructure is set up).
- They add extra time to jibs that don't need the persisted stuff (unless the restoration time is (and remains) negligible).

File: .circleci/config.yml (hunk 1 above)
Comment (filipesilva) [reply .circleci/config.yml]: Fixed

**Comment by filipesilva** [comment]
@gkalpak fixed the missing `install` references.

WRT to total workspace size, I'd like to highlight that `bazel_repository_cache` was already being cached, and restored, without workspaces. But #28928 erroneously did not persist it to the the workspace. That is a bug, and should be fixed.

As far as I can tell the restoration time of the workspace is less than the total setup time as it was before.
In the newly added examples in https://github.com/angular/angular/pull/28928#issue-255473932 the total workspace restoration is ~12 to 16s, whereas the total setup without workspaces was 27s to 51s.

I think there is some locality because [this job](https://circleci.com/gh/angular/angular/218903) was rerun and suddenly the workspace took 48s to download instead of 12s. But since cache download times for similar sizes are comparable, I think cache also has locality.

I think the current workspace is somewhat indiscriminate now, and has opportunities for optimization. For instance, we could not persist the git folder, and extract any git related information up front.

File: .circleci/config.yml [hunk 2]
```diff
@@ -136,16 +137,26 @@ jobs:
...
+      # Make the bazel directories and add a file to them if they don't exist already so that
+      # persist_to_workspace does not fail.
+      - run: |
+          if [ ! -d ~/bazel_repository_cache ]; then
+            mkdir ~/bazel_repository_cache
+            touch ~/bazel_repository_cache/MARKER
+          fi
       # Persist any changes at this point to be reused by further jobs.
+      # **NOTE 1 **: Folders persisted here should be kept in sync with `var_13: &attach_workspace`.
+      # **NOTE 2 **: To add new content to the workspace, always persist on the same root.
```
Comment (gkalpak) [inline .circleci/config.yml]: I don't think this is necessary, but is probably a good idea :smiley:

**Review by gkalpak — APPROVED** [review]

File: .circleci/config.yml (hunk 2 above)
Comment (filipesilva) [reply .circleci/config.yml]: I thought it wasn't either actually so initially I didn't change the root of `build-npm-packages`. Then the subsequent jobs couldn't find the folders. 

Then I thought maybe I'd have to attach the workspace at different roots. That didn't work either. Then I tried attaching multiple workspaces and I think only the last one got attached, but my memory on this one is fuzzy now. 

So I settled on "always add on the same root".

**Pull request closed without merging**